### PR TITLE
Fix #10656 - PolygonClipping: CLIPLINE produces NaN when line to clip is parallel to the maxX of the clip window

### LINKED ISSUE #10656: PolygonClipping: CLIPLINE produces NaN when line to clip is parallel to the maxX of the clip window
Issue overview
--------------

Here is an example:

```c++
    Real64 const minX = 2.0;
    Real64 const maxX = 8.0;
    Real64 const minY = 3.0;
    Real64 const maxY = 6.0;

    Real64 x0 = maxX;
    Real64 x1 = maxX;
    Real64 y0 = 4.5;
    Real64 y1 = 1.0;
    bool visible = false;
    bool rev = false;

    EXPECT_NO_THROW(CLIPLINE(x0, x1, y0, y1, maxX, minX, maxY, minY, visible, rev));

    // y1 is NaN now
```

Graphically, this looks like this:

image

Because dx is 0, This is a divide by zero on that line: https://github.com/NREL/EnergyPlus/blob/eb6665f392fc51eb6727ec5d8f7af21722941d79/src/EnergyPlus/SolarShading.cc#L3996

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version): all
 - Version of EnergyPlus (if using an intermediate build, include SHA): develop, eb6665f392fc51eb6727ec5d8f7af21722941d79
- Found in #10648

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [x] Ticket added to EnergyPlus Defect Complexity (Github Project)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

- Fixes #10656

Use the Liang-Barsky algorithm

* Loosely adapted from https://en.wikipedia.org/wiki/Liang%E2%80%93Barsky_algorithm
* Rewrote it until performance is optimal

I used microbenchmarking: https://github.com/jmarrec/CppBenchmarks/blob/94f0e1594ff17f9833a0547f0725d68c561311a1/bench_clipline.cpp

BM_CLIPLINE is the current E+ algorithm. BM_liang_barsky_clipper3 is my final version that I used here, and which is 15% faster than the original while being actually correct. in the edge cases

```
-------------------------------------------------------------------
Benchmark                         Time             CPU   Iterations
-------------------------------------------------------------------
BM_CLIPLINE                     567 ns          566 ns      1231826
BM_Clip2                        703 ns          703 ns       991672
BM_liang_barsky_clipper         588 ns          588 ns      1199750
BM_liang_barsky_clipper2        490 ns          490 ns      1430600
BM_liang_barsky_clipper3        482 ns          482 ns      145236
```

bench_clipline

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: tst/EnergyPlus/unit/SolarShading.unit.cc [hunk 1]
```diff
@@ -6204,3 +6204,280 @@ TEST_F(EnergyPlusFixture, SolarShadingTest_GetShadowingInputTest6)
...
+    Real64 y1 = 1.0;
+    bool visible = false;
+
+    EXPECT_NO_THROW(CLIPLINE(x0, x1, y0, y1, maxX, minX, maxY, minY, visible));
+
+    EXPECT_DOUBLE_EQ(maxX, x0);
+    EXPECT_DOUBLE_EQ(4.5, y0);
+    EXPECT_DOUBLE_EQ(maxX, x1);
+    EXPECT_DOUBLE_EQ(minY, y1); // This is NaN
+}
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/SolarShading.unit.cc:6227]: Simple test, was enough to produce a NaN for #10656

File: tst/EnergyPlus/unit/SolarShading.unit.cc [hunk 2]
```diff
@@ -6204,3 +6204,280 @@ TEST_F(EnergyPlusFixture, SolarShadingTest_GetShadowingInputTest6)
...
+    };
+    i = 0;
+    for (const auto &t : boundary_lines) {
+        ++i;
+        std::string const msg =
+            fmt::format("Boundary Line {}: From ({}, {}) to ({}, {})", i, t.line_ori.p0.x, t.line_ori.p0.y, t.line_ori.p1.x, t.line_ori.p1.y);
+        SCOPED_TRACE(msg);
+        testclipline(t);
+    }
+}
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/SolarShading.unit.cc]: Bunch more testing to ensure whatever algorithm I ended up with was actually correct.

I test from from one zone to another
```c++
    // We break the plane into 9 parts, where region (4) is the clipping region
    //
    //  6 │ 7 │ 8
    // ───────────
    //  3 │ 4 │ 5
    // ───────────
    //  0 │ 1 │ 2
```

And I also test on boundary lines, where the line is parallel (and on) the boundary of the clip window

File: src/EnergyPlus/SolarShading.hh [hunk 3]
```diff
@@ -136,6 +136,8 @@ namespace SolarShading {
 
     void CLIP(EnergyPlusData &state, int const NVT, Array1D<Real64> &XVT, Array1D<Real64> &YVT, Array1D<Real64> &ZVT);
 
+    void CLIPLINE(Real64 &x0, Real64 &x1, Real64 &y0, Real64 &y1, Real64 maxX, Real64 minX, Real64 maxY, Real64 minY, bool &visible);
```
Comment (jmarrec) [inline src/EnergyPlus/SolarShading.hh:139]: I expose it so I can test it. I'm also handling the rev inside the routine, makes no sense to leave the caller to deal with it

File: src/EnergyPlus/SolarShading.cc [hunk 4]
```diff
@@ -3868,162 +3869,99 @@ inline bool d_eq(Real64 a, Real64 b)
     return std::abs(a - b) < 2.0;
 }
 
-void CLIPLINE(Real64 &x1, Real64 &x2, Real64 &y1, Real64 &y2, Real64 maxX, Real64 minX, Real64 maxY, Real64 minY, bool &visible, bool &rev)
+void CLIPLINE(Real64 &x0, Real64 &x1, Real64 &y0, Real64 &y1, Real64 maxX, Real64 minX, Real64 maxY, Real64 minY, bool &visible)
```
Comment (Myoldmopar) [inline src/EnergyPlus/SolarShading.cc:3872]: This is fantastic.  So much cleaner!

File: tst/EnergyPlus/unit/SolarShading.unit.cc [hunk 5]
```diff
@@ -6204,3 +6204,285 @@ TEST_F(EnergyPlusFixture, SolarShadingTest_GetShadowingInputTest6)
...
+        // From 8 to 5
+        TestCase{Line{Point{greater_x, greater_y}, Point{greater_x, center_y}}, false},
+    }};
+
+    size_t i = 0;
+    for (const auto &t : test_cases) {
+        ++i;
+        std::string const msg =
+            fmt::format("test_case {}: From ({}, {}) to ({}, {})", i, t.line_ori.p0.x, t.line_ori.p0.y, t.line_ori.p1.x, t.line_ori.p1.y);
+        SCOPED_TRACE(msg);
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/SolarShading.unit.cc:6450]: Does this always emit the current position?  I will see it locally if so, just asking to remind myself to look I guess.

**Review by Myoldmopar — APPROVED** [review]
I'll mark it approved but actually I will remove the SCOPED_TRACE if that actually adds outputs during the test runs.

File: tst/EnergyPlus/unit/SolarShading.unit.cc (hunk 5 above)
Comment (jmarrec) [reply tst/EnergyPlus/unit/SolarShading.unit.cc:6450]: SCOPED_TRACE will emit everytime another assertion fails. So it emits nothing if the test passes, but if it something goes wrong in the lambda with the EXPECT statements, you know which iteration of the loop triggered it.

Comment (Myoldmopar) [reply tst/EnergyPlus/unit/SolarShading.unit.cc:6450]: OK, that's great. I thought it was just emitting info by calling the macro itself.  I now see it is just registering the extra data to report in case of a failure.  This is great, thanks.

**Review by Myoldmopar — APPROVED** [review]
OK, let's go for it.

**Comment by Myoldmopar** [comment]
I pulled develop, and made sure it was still happy.  I'll get this merged so that I can resolve conflicts in #10578.  Thanks @jmarrec , merging.
